### ISSUE TITLE
Use multiprocessing to parallely process PDF pages

### ISSUE THREAD

**Issue by vinayak-mehta** [camelot-dev/camelot#20]
<pre>
>>> camelot.read_pdf('filename.pdf', pages='all', parallel=True)
</pre>

We could try and use all cores present on the machine using multiprocessing. More ideas are welcome.

**Comment by sathk882** [comment]
Hi @vinayak-mehta ,

Even I thought of implementing this. [dramatiq](https://github.com/Bogdanp/dramatiq) or [celery](https://github.com/celery/celery) are my suggestions for asynchronous processing of pages.

**Comment by jontis** [comment]
I'm doing this with dask but it's chosen out of habit.

**Comment by selcukusta** [comment]
Is there any improvement in there? I have a file that has only one page. The page has a table (25 rows x 13 columns). Each cell has `read_pdf` function takes 10 seconds after that `to_excel` takes only 100-150 ms. I'm thinking about 10 seconds is too long, am I wrong?

**Comment by mr-bjerre** [comment]
@selcukusta I think this is more about running multiple pages at the same time rather than speeding up the extraction of a single page.

But does anyone have a solution for multiple pages in parallel?

**Comment by vinayak-mehta** [comment]
> @selcukusta I think this is more about running multiple pages at the same time rather than speeding up the extraction of a single page.

Yes!

> But does anyone have a solution for multiple pages in parallel?

Using multiprocessing, we should be able to distribute multiple pages on all cores, processing them in parallel.

**Comment by mr-bjerre** [comment]
I get this though

    objc[53475]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.

**Comment by selcukusta** [comment]
> @selcukusta I think this is more about running multiple pages at the same time rather than speeding up the extraction of a single page.
> 
> But does anyone have a solution for multiple pages in parallel?

Yeah, I know. Actually it's related with [that](https://github.com/atlanhq/camelot/issues/301) but the issue was closed and referenced to it.

**Comment by rawsh-bt** [comment]
Does anyone have an update? I've tried inheriting PageHandler and making parsing multithreaded / multicore, and multi threading processing multiple pdfs, but I'm running into a ghostscript error (seems like it's not thread safe?)

**Comment by phoewass** [comment]
I did implement a multi-threading layer above camelot.read_pdf using **multiprocessing** library.
I faced a couple of pitfalls doing it, so I can help on this if I may.

**Comment by vinayak-mehta** [comment]
@phoewass That would be awesome if you're still interested!

**Comment by RickyGunawan09** [comment]
can anyone tell me how to use multiprocess in camelot ? or did this issues still on progress ?

**Comment by phoewass** [comment]
Hi all. Sorry it took me a while to publish the PR while the code was already available.
Now the PR is there to be reviewed, I'm looking forward for your feedback.

**Comment by vinayak-mehta** [comment]
👀

**Comment by Siddharth1India** [comment]
Any update on this? My PDFs are 100s of pages and I can really use this feature.

**Comment by mlbrothers** [comment]
@phoewass @vinayak-mehta is this feature part of library now? If not, is there any way I can utilize multiprocessing to read multipage PDF?

**Comment by deepakagrawal** [comment]
Any updates on this features?

**Comment by bosd** [comment]
Hey!

 As [camelot is dead](https://github.com/camelot-dev/camelot/issues/343), we try to build a maintained fork at [`pypdf_table_extraction`](https://github.com/py-pdf/pypdf_table_extraction).

There is a discussion about this in: 

https://github.com/py-pdf/pypdf_table_extraction/discussions/8#discussioncomment-8617535

**Comment by bosd** [comment]
Closing as implemented — `read_pdf(..., parallel=True)` has shipped for a while now, and #712 added the `cpu_count=N` cap so users can limit worker count. The map-reduce over pages happens inside `playa.open(...).pages[...].map(...)`. If there's a specific multiprocessing case still not supported (e.g. process pool reuse across calls), opening a fresh focused issue would be the way to go.
